### DISCUSSION TITLE
How can i use rate limit?

### DISCUSSION THREAD

**Discussion by B4TT3RY** [tokio-rs/axum#987] [Q&A] [answered]
I want to apply a rate limit to my middleware.
I tried using [RateLimitLayer](https://docs.rs/tower/latest/tower/limit/rate/struct.RateLimitLayer.html) on `tower (v0.4.12)`, but I get a compile error.
How can I use it?

**Comment by davidpdrsn** [comment, marked as answer]
What's the error?

**Reply by B4TT3RY** [reply]
Cargo.toml
`tower = { version = "0.4.12", features = ["limit"] }`

main.rs
```rs
let app = Router::new()
  .route(<My Router>)
  .layer(RateLimitLayer::new(5, Duration::from_secs(1)));
```

run `cargo check`
```
error[E0277]: the trait bound `RateLimit<Route<_>>: Clone` is not satisfied                                                                                                                                                      
   --> src/main.rs:96:10
    |
96  |         .layer(RateLimitLayer::new(5, Duration::from_secs(1)))
    |          ^^^^^ the trait `Clone` is not implemented for `RateLimit<Route<_>>`
    |
note: required by a bound in `Router::<B>::layer`
   --> /home/b4tt3ry/.cargo/registry/src/github.com-1ecc6299db9ec823/axum-0.5.4/src/routing/mod.rs:295:15
    |
295 |             + Clone
    |               ^^^^^ required by this bound in `Router::<B>::layer`
```

**Reply by davidpdrsn** [reply]
The issue is that rate limited services aren't cloneable, which kinda makes sense. If they were you'd be able to just clone them to fake being able to handle more load.

The solution is to use `tower::buffer::BufferLayer` to make the service cloneable by running it on a background task and sending requests to it via channel.

Something like this:

```rust
use axum::{error_handling::HandleErrorLayer, http::StatusCode, routing::get, BoxError, Router};
use std::{net::SocketAddr, time::Duration};
use tower::{buffer::BufferLayer, limit::RateLimitLayer, ServiceBuilder};

#[tokio::main]
async fn main() {
    let app = Router::new().route("/", get(|| async {})).layer(
        ServiceBuilder::new()
            .layer(HandleErrorLayer::new(|err: BoxError| async move {
                (
                    StatusCode::INTERNAL_SERVER_ERROR,
                    format!("Unhandled error: {}", err),
                )
            }))
            .layer(BufferLayer::new(1024))
            .layer(RateLimitLayer::new(5, Duration::from_secs(1))),
    );

    let addr = SocketAddr::from(([127, 0, 0, 1], 3000));
    axum::Server::bind(&addr)
        .serve(app.into_make_service())
        .await
        .unwrap();
}
```

**Reply by B4TT3RY** [reply]
Thank you very much!

**Reply by davidpdrsn** [reply]
np :)

**Reply by kimbauters** [reply]
Another big thanks! Most made sense, but that `HandleErrorLayer` was a tricky one (in hindsight obvious, of course, sigh).

**Reply by eboody** [reply]
thanks for this!

**Reply by shnartho** [reply]
I have been trying to implement rate limit for last few days. No AI bot could help, only this thread solved the issue. Thank you.

**Reply by dayvejones** [reply]
https://github.com/benwis/tower-governor

Working solution.

**Reply by Maciej-Suchanski** [reply]
thank you for working out of the box solution ❤️

**Reply by sbag13** [reply]
the above solution actually compiles but does not limit the rate

**Reply by nivek3** [reply]
https://crates.io/crates/tower-rate-limiter
